// @mentra/types/dist/enums.js
"use strict";
/**
 * @mentra/types - Shared enums for MentraOS
 * These are runtime values (not pure types)
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeviceTypes = exports.HardwareRequirementLevel = exports.HardwareType = void 0;
/**
 * Hardware component types that apps can require
 */
var HardwareType;
(function (HardwareType) {
    HardwareType["CAMERA"] = "CAMERA";
    HardwareType["DISPLAY"] = "DISPLAY";
    HardwareType["MICROPHONE"] = "MICROPHONE";
    HardwareType["SPEAKER"] = "SPEAKER";
    HardwareType["IMU"] = "IMU";
    HardwareType["BUTTON"] = "BUTTON";
    HardwareType["LIGHT"] = "LIGHT";
    HardwareType["WIFI"] = "WIFI";
})(HardwareType || (exports.HardwareType = HardwareType = {}));
/**
 * Levels of hardware requirements
 */
var HardwareRequirementLevel;
(function (HardwareRequirementLevel) {
    HardwareRequirementLevel["REQUIRED"] = "REQUIRED";
    HardwareRequirementLevel["OPTIONAL"] = "OPTIONAL";
})(HardwareRequirementLevel || (exports.HardwareRequirementLevel = HardwareRequirementLevel = {}));
var DeviceTypes;
(function (DeviceTypes) {
    DeviceTypes["SIMULATED"] = "Simulated Glasses";
    DeviceTypes["G1"] = "Even Realities G1";
    DeviceTypes["LIVE"] = "Mentra Live";
    DeviceTypes["MACH1"] = "Mentra Mach1";
    DeviceTypes["Z100"] = "Vuzix Z100";
    DeviceTypes["NEX"] = "Mentra Nex";
    DeviceTypes["FRAME"] = "Brilliant Frame";
})(DeviceTypes || (exports.DeviceTypes = DeviceTypes = {}));
