// @mentra/types/dist/hardware.js
"use strict";
/**
 * @mentra/types - Hardware capability types
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.mentraDisplay = exports.vuzixZ100 = exports.mentraLive = exports.evenRealitiesG2 = exports.evenRealitiesG1 = exports.simulatedGlasses = exports.getModelCapabilities = exports.HARDWARE_CAPABILITIES = void 0;
const even_realities_g1_1 = require("./capabilities/even-realities-g1");
Object.defineProperty(exports, "evenRealitiesG1", { enumerable: true, get: function () { return even_realities_g1_1.evenRealitiesG1; } });
const even_realities_g2_1 = require("./capabilities/even-realities-g2");
Object.defineProperty(exports, "evenRealitiesG2", { enumerable: true, get: function () { return even_realities_g2_1.evenRealitiesG2; } });
const mentra_display_1 = require("./capabilities/mentra-display");
Object.defineProperty(exports, "mentraDisplay", { enumerable: true, get: function () { return mentra_display_1.mentraDisplay; } });
const mentra_live_1 = require("./capabilities/mentra-live");
Object.defineProperty(exports, "mentraLive", { enumerable: true, get: function () { return mentra_live_1.mentraLive; } });
const simulated_glasses_1 = require("./capabilities/simulated-glasses");
Object.defineProperty(exports, "simulatedGlasses", { enumerable: true, get: function () { return simulated_glasses_1.simulatedGlasses; } });
const vuzix_z100_1 = require("./capabilities/vuzix-z100");
Object.defineProperty(exports, "vuzixZ100", { enumerable: true, get: function () { return vuzix_z100_1.vuzixZ100; } });
const none_1 = require("./capabilities/none");
const enums_1 = require("./enums");
/**
 * Hardware capability profiles for supported glasses models
 * Key: model_name string (e.g., "Even Realities G1", "Mentra Live")
 * Value: Capabilities object defining device features
 */
exports.HARDWARE_CAPABILITIES = {
    [even_realities_g1_1.evenRealitiesG1.modelName]: even_realities_g1_1.evenRealitiesG1,
    [even_realities_g2_1.evenRealitiesG2.modelName]: even_realities_g2_1.evenRealitiesG2,
    [mentra_display_1.mentraDisplay.modelName]: mentra_display_1.mentraDisplay,
    [mentra_live_1.mentraLive.modelName]: mentra_live_1.mentraLive,
    [simulated_glasses_1.simulatedGlasses.modelName]: simulated_glasses_1.simulatedGlasses,
    [vuzix_z100_1.vuzixZ100.modelName]: vuzix_z100_1.vuzixZ100,
    [enums_1.DeviceTypes.MACH1]: vuzix_z100_1.vuzixZ100, // Mach1 uses same Vuzix Ultralite hardware as Z100
    [none_1.none.modelName]: none_1.none,
};
const getModelCapabilities = (deviceType) => {
    const modelName = deviceType;
    if (!exports.HARDWARE_CAPABILITIES[modelName]) {
        return exports.HARDWARE_CAPABILITIES[enums_1.DeviceTypes.NONE];
    }
    return exports.HARDWARE_CAPABILITIES[modelName];
};
exports.getModelCapabilities = getModelCapabilities;
